src/dune/simple_rules.ml

```diff
@@ -35,10 +56,43 @@ let check_filename =
       Path.as_in_build_dir_exn p
     | Dir p -> not_in_dir ~error_loc (Path.to_string p)
 
+type rule_kind =
+  | Alias_only of Alias.Name.t
+  | Alias_with_targets of Alias.Name.t * Path.Build.t
+  | No_alias
+
+let rule_kind ~(rule : Rule.t) ~action =
+  match rule.alias with
+  | None -> No_alias
+  | Some alias -> (
+    match Build.targets action |> Path.Build.Set.choose with
+    | None -> Alias_only alias
+    | Some target -> Alias_with_targets (alias, target) )
+
+let add_user_rule sctx ~dir ~(rule : Rule.t) ~action ~expander =
+  SC.add_rule_get_targets
+    sctx
+    (* user rules may have extra requirements, in which case they will be
+       specified as a part of rule.deps, which will be correctly taken care of
+       by the build description *)
+    ~sandbox:Sandbox_config.no_special_requirements ~dir ~mode:rule.mode
+    ~loc:rule.loc
+    ~locks:(interpret_locks ~expander rule.locks)
+    action
+
 let user_rule sctx ?extra_bindings ~dir ~expander (rule : Rule.t) =
   match Expander.eval_blang expander rule.enabled_if with
-  | false -> Path.Build.Set.empty
-  | true ->
+  | false ->
+    Option.iter rule.alias ~f:(fun name ->
```

Comment:
What's the idea here? Why are we adding something at all if the rule is disabled?

Reply:
The idea is that we want `$ dune build @foo` to work regardless if the rule is enabled or not. Consider `$ dune build @runtest` where a test might potentially be disabled. We don't want to get an error like "this alias is empty"

Reply:
Ok, fair enough, but you're still passing rule.deps. Do you mean empty deps?

Reply:
Hmm, good point. I'm not sure why I did it that way but it's wrong. The deps are no longer interpreted.